RFC pyodide governance and decision-making process (#1229)
iodide-project/pyodide: docs/project/governance.md

the continued development of the project through ongoing engagement with the
community. They have shown they can be trusted to maintain pyodide with
care. Being a core developer allows contributors to more easily carry on
with their project related activities by giving them direct access to the
project’s repository and is represented as being a member of the core team on the
pyodide [GitHub organization](https://github.com/orgs/pyodide/teams/core/members).
Core developers are expected to review code
contributions, can merge approved pull requests, can cast votes for and against
merging a pull-request, and can make decisions about major changes to the
API (all contributors are welcome to participate in the discussion).

New core developers can be nominated by any existing core developers. Once they
have been nominated, there will be a vote by the current core developers.
Voting on new core developers is one of the few activities that takes place on
the project's private communication channels. While it is expected that most votes
will be unanimous, a two-thirds majority of the cast votes is enough. The vote

> needs to be open for at least 1 week.

Reviewer: "one week"?

Core developers that have not contributed to the project (commits or GitHub

> comments) in the past 2 years will be asked if they want to become emeritus

Reviewer: "two years"?

core developers and recant their commit and voting rights until they become
active again.

## Decision Making Process

Decisions about the future of the project are made through discussion with all
members of the community. All non-sensitive project management discussion takes
place on the project contributors' [issue
tracker](https://github.com/pyodide/pyodide/issues) and on [Github
discussion](https://github.com/pyodide/pyodide/discussions).
Occasionally, sensitive discussion occurs on a private communication channels.

pyodide uses a "consensus seeking" process for making decisions. The group
tries to find a resolution that has no open objections among core developers.
At any point during the discussion, any core-developer can call for a vote,

> which will conclude two weeks from the call for the vote. This is what we

Reviewer: here, it is already "two weeks" :+1:

hereafter may refer to as “the decision making process”.

Decisions (in addition to adding core developers as above)
are made according to the following rules:

* **Maintenance changes**, include for
  instance improving the wording in the documentation, updating CI or
  dependencies.  Core developers are expected to give “reasonable time” to
  others to give their opinion on the pull request if they’re not confident
  others would agree. If no review is received within this time, the Pull

>   Request can be merged.

Reviewer: I'm not happy with these two sentences, because they're a little bit confusing.

> Core developers are expected to give “reasonable time” to
  others to give their opinion on the pull request if they’re not confident
  others would agree.

Might this be changed into `Core developers are expected to give “reasonable time” to
  others to give their opinion on the Pull Request in case they’re not confident
  that others would agree.` ?

> If no review is received within this time, the Pull Request can be merged.

Change this into `If no further review or blocking comment on the Pull Request is received within this time, it be merged.`?

# pyodide governance and decision-making

The purpose of this document is to formalize the governance process used by the
pyodide project, to clarify how decisions are made and how the various

> elements of our community interact.

Reviewer 2: ```suggestion
members of our community interact.
```

This document establishes a decision-making structure that takes into account
feedback from all members of the community and strives to find consensus, while
avoiding any deadlocks.

Anyone with an
interest in the project can join the community, contribute to the project
design and participate in the decision making process. This document describes

> how that participation takes place and how to set about earning merit within
> the project community.

Reviewer 2: ```suggestion
how to participate and earn merit in the pyodide community.
```

## Roles And Responsibilities

### Contributors

Contributors are community members who contribute in concrete ways to the
project. Anyone can become a contributor, and contributions can take many forms
– not only code – as detailed in the {ref}`how_to_contibute`.

### Community members team

The community members team is composed of community members who have permission on

> Github to label and close issues. Their work is

Author: Not sure about the name for this team. iodode has "community members team", the other name could be "triage team".

Another side effect from being in this team is that it would be possible to request review on PRs from its members via Github UI. That feature doesn't work for users not part of the org.

Reviewer 2: I think "community members team" sounds friendlier, "triage team" sounds gory and like it requires medical training.

crucial to improve the communication in the project.

> Similarly to what has been decided in the [CPython project](
> https://devguide.python.org/triaging/#becoming-a-member-of-the-python-triage-team),
> any contributor may become a member of this team, after
> showing some continuity in participating in pyodide
> development (with pull requests and reviews).

Reviewer 2: ```suggestion
After participating in pyodide development with pull requests and reviews for a period of time, any contributor may become a member of the team.
The process for adding team members is modeled on the [CPython project](
https://devguide.python.org/triaging/#becoming-a-member-of-the-python-triage-team).
```

Reviewer 2: Do we really want to focus on PRs and reviews here? Seems like we should also emphasize helping users in various ways (answering questions on gitter, responding to issues explaining how to do stuff).

place on the project contributors' [issue
tracker](https://github.com/pyodide/pyodide/issues) and on [Github
discussion](https://github.com/pyodide/pyodide/discussions).
Occasionally, sensitive discussion occurs on a private communication channels.

pyodide uses a "consensus seeking" process for making decisions. The group
tries to find a resolution that has no open objections among core developers.
At any point during the discussion, any core-developer can call for a vote,
which will conclude two weeks from the call for the vote. This is what we
hereafter may refer to as “the decision making process”.

Decisions (in addition to adding core developers as above)
are made according to the following rules:

* **Minor Documentation changes and minor build setup changes**, include for
  instance improving the wording in the documentation, updating CI or
  dependencies.  Core developers are expected to give “reasonable time” to
  others to give their opinion on the pull request if they’re not confident
  others would agree. If no review is received within this time, the Pull
  Request can be merged.

* **Code changes impacting user facing APIs, or having backward compatibility

>   implications** require +1 by a core developers, no -1 by a core developer

Author: What constitutes a backward incompatible change in pyodide is still a bit unclear. In the sense that updating emsdk could also break some backward compatibility. I imagine this means not breaking use user code that's working for some version of pyodide, excluding potential breakage due to python packages updates.

Reviewer 2: I'm not really sure what the text here means. Does it mean that for changes to behavior / user facing API, at least one core dev other than the one making the PR should review the PR and approve it before merging? This seems to match the current procedure?

Author: Yes, that's what it mean and  what we are mostly doing now. I'll replace "+1" by "a review and an approval (+1 vote)" and "-1" by "objection with a -1 vote" if it helps.

# pyodide governance and decision-making

The purpose of this document is to formalize the governance process used by the
pyodide project, to clarify how decisions are made and how the various
elements of our community interact.
This document establishes a decision-making structure that takes into account
feedback from all members of the community and strives to find consensus, while
avoiding any deadlocks.

Anyone with an
interest in the project can join the community, contribute to the project
design and participate in the decision making process. This document describes
how that participation takes place and how to set about earning merit within
the project community.

## Roles And Responsibilities

### Contributors

Contributors are community members who contribute in concrete ways to the
project. Anyone can become a contributor, and contributions can take many forms

> – not only code – as detailed in the {ref}`how_to_contibute`.

Reviewer 2: Would be nice to explicitly mention e.g., answering user questions as a way to contribute here.

Author: Absolutely, but I think different ways of how to contribute, such as answering user questions, reviewing PRs should be listed in https://pyodide.readthedocs.io/en/latest/development/contributing.html not in the governance. That page needs re-organizing anyway.

For instance an example of such section in https://scikit-learn.org/stable/developers/contributing.html#ways-to-contribute

Reviewer 2: I think giving a single example like "answering user questions" here might help indicate that we mean "contribute" in a very inclusive sense.

any contributor may become a member of this team, after
showing some continuity in participating in pyodide
development (with pull requests and reviews).
Any core developer is welcome to propose a pyodide contributor to join the
community members team. Other core developers are then consulted: while it is expected
that most acceptances will be unanimous, a two-thirds majority is enough.

### Core developers

Core developers are community members who have shown that they are dedicated to
the continued development of the project through ongoing engagement with the
community. They have shown they can be trusted to maintain pyodide with
care. Being a core developer allows contributors to more easily carry on
with their project related activities by giving them direct access to the
project’s repository and is represented as being a member of the core team on the
pyodide [GitHub organization](https://github.com/orgs/pyodide/teams/core/members).
Core developers are expected to review code
contributions, can merge approved pull requests, can cast votes for and against
merging a pull-request, and can be involved in deciding major changes to the

> API.

Reviewer 2: Might be nice to clarify the wording here, but I'm not quite clear how. Everyone can be involved in the consensus process? We explicitly want to include users who haven't contributed code in discussions about major API changes. If a lot of users don't like it, we might consider not doing it. I guess here it's intended to mean that the final up-or-down judgment calls in cases of contention go to core devs?

New core developers can be nominated by any existing core developers. Once they
have been nominated, there will be a vote by the current core developers.
Voting on new core developers is one of the few activities that takes place on

> the project's private communication channels. While it is expected that most votes

Reviewer 2: > private communication channels 

Is there a plan to set up private communication channels? I assume we don't currently have these.

Author: Yes, we need to decide which ones. One possibility is ask for creation of a pyodide@python.org mailing list. The other is to use the comment system specific to Github teams, but that would tend to get lost in all the other GH notifications I think.

will be unanimous, a two-thirds majority of the cast votes is enough. The vote
needs to be open for at least 1 week.

Core developers that have not contributed to the project (commits or GitHub
comments) in the past 2 years will be asked if they want to become emeritus
core developers and recant their commit and voting rights until they become
active again.

## Decision Making Process

Decisions about the future of the project are made through discussion with all
members of the community. All non-sensitive project management discussion takes
place on the project contributors' [issue
tracker](https://github.com/pyodide/pyodide/issues) and on [Github
discussion](https://github.com/pyodide/pyodide/discussions).
Occasionally, sensitive discussion occurs on a private communication channels.

pyodide uses a "consensus seeking" process for making decisions. The group
tries to find a resolution that has no open objections among core developers.
At any point during the discussion, any core-developer can call for a vote,
which will conclude two weeks from the call for the vote. This is what we

> hereafter may refer to as “the decision making process”.

Reviewer 2: It's a little unclear to me what scenarios would require a formal vote, is it just for any issue that leads to a significant amount of contention?

Author: The only scenario where a formal note needs to happen in the current document is for adding new maintainers. All other points should ideally reach a consensus by discussion, and only fall back on voting if agreement cannot be reached and a decision needs to be taken.

Decisions (in addition to adding core developers as above)
are made according to the following rules:

> * **Minor Documentation changes and minor build setup changes**, include for

Reviewer 2: Maybe these could be called "maintenance changes"

  instance improving the wording in the documentation, updating CI or
  dependencies.  Core developers are expected to give “reasonable time” to
  others to give their opinion on the pull request if they’re not confident
  others would agree. If no review is received within this time, the Pull
  Request can be merged.

* **Code changes impacting user facing APIs, or having backward compatibility

>   implications** require a review and an approval (+1 vote) by a core
>   developer, no objections with -1 vote by a core developer (lazy consensus).
>   This process happens on the pull-request page.

Reviewer 2: ```suggestion
  implications** require review and approval by a core
  developer and no objections raised by any core developer (lazy consensus).
  This process happens on the pull-request page.
```
Maybe say again that there should be reasonable time given and also suggest @notifying community members that one expects to have a stake in the matter? Maybe also clarify that it needs to be reviewed by a core developer different from the one who opened the PR?

I think it's confusing to say +1 vote / -1 vote because it's not clear to me that there is any formal way to do that / any formal difference between "I object to this" and "-1"?

Author: > suggest notifying community members that one expects to have a stake in the matter?

I think it's up to the contributor to decide about mentioning or not. Generally maintainers are often following the repo and will see the PR happen. Whether they will respond or not depends on their availability, interest in the particular topic etc and so systematically mentioning is nor really useful if you follow all issues anyway.

> Maybe also clarify that it needs to be reviewed by a core developer different from the one who opened the PR?

You need an approval to merge those PR and you can't approve your own PRs in the Github UI so I think that excludes this case.

Reviewer 2: > you can't approve your own PRs in the Github UI

Oh okay, I hadn't tried that yet.

Decisions (in addition to adding core developers as above)
are made according to the following rules:

* **Maintenance changes**, include for
  instance improving the wording in the documentation, updating CI or
  dependencies.  Core developers are expected to give “reasonable time” to
  others to give their opinion on the pull request if they’re not confident
  others would agree. If no review is received within this time, the Pull
  Request can be merged.

> * **Code changes in general, and expecially those impacting user facing APIs**,

Reviewer 2: Spelling: "expecially" ==> "especially"
